File: src/com/crossbowffs/quotelock/backup/ProgressCallback.java
package com.crossbowffs.quotelock.backup;

import android.os.Looper;

import androidx.annotation.MainThread;

import com.crossbowffs.quotelock.utils.AppExecutors;

/**
 * @author Yubyf
 * @date 2021/6/13.
 */
public interface ProgressCallback {

    @MainThread
    void inProcessing(String message);

    @MainThread
    void success();

    @MainThread
    void failure(String message);

    default void safeInProcessing(String message) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            AppExecutors.getInstance().mainThread().execute(() -> inProcessing(message));
            return;
        }
        inProcessing(message);
    }

    default void safeSuccess() {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            AppExecutors.getInstance().mainThread().execute(this::success);
            return;
        }
        success();
    }


    default void safeFailure(String message) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            AppExecutors.getInstance().mainThread().execute(() -> failure(message));
            return;
        }
        failure(message);
    }
}

File: app/src/main/java/com/crossbowffs/quotelock/backup/ProgressCallback.kt
package com.crossbowffs.quotelock.backup

import android.os.Looper
import androidx.annotation.MainThread
import com.crossbowffs.quotelock.utils.mainScope
import kotlinx.coroutines.launch

/**
 * @author Yubyf
 * @date 2021/6/13.
 */
interface ProgressCallback {
    @MainThread
    fun inProcessing(message: String?)

    @MainThread
    fun success(message: String?)

    @MainThread
    fun failure(message: String?)

    fun safeInProcessing(message: String?) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            mainScope.launch { inProcessing(message) }
            return
        }
        inProcessing(message)
    }

    fun safeSuccess(message: String? = "") {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            mainScope.launch { success(message) }
            return
        }
        success(message)
    }

    fun safeFailure(message: String?) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            mainScope.launch { failure(message) }
            return
        }
        failure(message)
    }
}